### PR TITLE
Scanner first draft

### PR DESCRIPTION
**Pull request by j-hui**
Scanner should be mostly complete. In sum, this what we support:

- all keywords (i think)
- all operators (i think)
- string literals
- int literals
- boolean literals
- simple float literals
- identifiers
- 🔥 nested comments 🔥
- line numbers 🔑💯👌

todos:
- [x] expand syntax for floats?
- [x] figure out how these exceptions work
- [x] do we want all these boolean operators?
- [x] do we want all these bitwise operators?
- [ ] check keywords
- [x] figure out how this `unescape` function works
- [x] have a `Makefile` that isn't a dumbass hardcoded piece of shit

Referenced parsers from Funk and Dice

### CODE REVIEW

File: src/frontend/scanner.mll [hunk 1]
```diff
@@ -0,0 +1,139 @@
+{
+  open Parser
+  let lineno = ref 1
+  let depth = ref 0
+  let filename = ref "" (* what do with this *)
```
Comment (Contentmaudlin) [inline src/frontend/scanner.mll:5]: what even is this?

File: src/frontend/scanner.mll [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+| "*="    { MUL_ASN }
+| "/="    { DIV_ASN }
+| "%="    { MOD_ASN }
+
+(* logical operators *)
+| "&&"    { LOG_AND }
+| "||"    { LOG_OR }
+| "!"     { LOG_NOT }
+
+(* do we want to do this?
```
Comment (Contentmaudlin) [inline src/frontend/scanner.mll:73]: honestly no

File: src/frontend/scanner.mll [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+(* ye good olde *)
+| eof { EOF }
+| _ as e    { raise (Exceptions.IllegalCharacter(!filename, illegal, !lineno)) }
+
+(* comments *)
+(* does not support nested comments yet *)
+and comment = parse
+  "/*"      { incr depth; comment lexbuf }
+| "*/"      { decr depth; if depth > 0 then token lexbuf else token lexbuf }
```
Comment (Contentmaudlin) [inline src/frontend/scanner.mll]: can you run me through the logic here? How is "depth" used?  I assume you're using it for nested comments?

File: src/frontend/scanner.mll (hunk 1 above)
Comment (j-hui) [reply src/frontend/scanner.mll:5]: These are nice nice mutable variables

It's so our scanner can report syntax errors with line numbers.

File: src/frontend/scanner.mll (hunk 3 above)
Comment (j-hui) [reply src/frontend/scanner.mll]: Yeah so basically

f u c k D F A s

This is cheating and implementing a pushdown automaton to match `/*`s and `*-`s

Comment (Contentmaudlin) [reply src/frontend/scanner.mll]: So basically allows nested comments to happen or no?

File: src/frontend/scanner.mll (hunk 1 above)
Comment (Contentmaudlin) [reply src/frontend/scanner.mll:5]: Yeah but what is filename

Comment (j-hui) [reply src/frontend/scanner.mll:5]: no fucking clue how that's done yet, will find out. but this is done in Dice

File: src/frontend/scanner.mll (hunk 2 above)
Comment (j-hui) [reply src/frontend/scanner.mll:73]: yeah but i wnt to defer to @andylhxu (as our language guru) and @cerrno  (as our resident graphics guy aka primary consumer) to tell us whether these are needed.

File: src/frontend/scanner.mll (hunk 3 above)
Comment (j-hui) [reply src/frontend/scanner.mll]: yes, nested comments are a go 🏋️

File: src/frontend/scanner.mll (hunk 2 above)
Comment (lucas-schuermann) [reply src/frontend/scanner.mll:73]: No, they're not needed. We only need boolean logic for conditionals and then support for scalars

**Comment by lucas-schuermann** [comment]
Re: TODOs, yes we want all boolean operators for more complex conditionals. I think that it would severely limit our scope if we didn't have them.

File: src/frontend/scanner.mll [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+(* ye good olde *)
+| eof       { EOF }
+
+
+(* The Reign of Error *)
+| '"'       { raise (Exceptions.UnmatchedQuotation(!lineno)) }
+| _ as e    { raise (Exceptions.IllegalCharacter(!filename, e, !lineno)) }
+
+
+(* comments *)
```
Comment (lucas-schuermann) [inline src/frontend/scanner.mll:138]: Does this check for intermediate lines starting with *? As in
```
/*
* <--- 
*/

Comment (lucas-schuermann) [reply src/frontend/scanner.mll:138]: Or am I an idiot?

**Comment by lucas-schuermann** [comment]
Re: makefile, fine for now--we can just file a GitHub issue and make sure that it is changed later

**Review by lucas-schuermann — APPROVED** [review]
LGTM, merging

File: src/frontend/scanner.mll (hunk 4 above)
Comment (j-hui) [reply src/frontend/scanner.mll:138]: this doesn't, but i realised that maybe it's not such a good idea to do that. I can add that in if we really want, but I'm starting to wonder if it's too much of a syntactic restriction?

**Comment by lucas-schuermann** [comment]
Yeah I mean I'm fine with not having it

**Pull request merged**